### PR TITLE
BUGFIX: PostgreSQL compatibility for `Neos.Neos` - no graph

### PR DESCRIPTION
**Pull request by mhsdesign**
Carefully extracted from #5751 and reduced to the essential changes to ease reviewing

The CI tests now postgreSQL compatibility by running all tables in postgres except the content graph which runs in mariadb

Needs followup change for 9.1 which turns the new #5542 projection postgresql compatible -> https://github.com/neos/neos-development-collection/pull/5805

TODO:
- [x] add composer script to run m*sql vs postgresql tests, see `match` in github ci
- [x] remove obsolete `@contentrepository` and `@adapters=` tags in behat -> #5806 
- [x] ~review `AbstractSubscriptionEngineTestCase::resetDatabase` -> simplify with https://github.com/neos/neos-development-collection/pull/5802~ -> not easy
- [x] reduce ci runs to a sane default currently we test mysql / mariadb / postgresql version with different php version 8.2-8.5 but it might make more sense to rather test the minimum supported mysql with the lowest php and maximum mysql with newest php instead all combinations
  - like with testing mysql 8.0 we want to run postgreSQL only with php 8.4 in the matrix.

Later
- [ ] document min postgresql version? 16 for `any_value`?

**Upgrade instructions**

**Review instructions**

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the correct branch:
  - If it's a bugfix, use the [lowest maintained branch which has the bug](https://www.neos.io/features/release-roadmap.html)
  - If it's a non-breaking feature, use the branch of the next version (might be either minor or major)
  - If it's a breaking feature it should typically go into the next major version
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [x] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Review by skurfuerst — APPROVED** [review]
looks all good to me - checked it all again :) Tremendous work <3 <3

File: Neos.ContentRepository.Dbal/Classes/DbalSchemaDiff.php [hunk 1]
```diff
@@ -36,6 +37,14 @@ public static function determineRequiredSqlStatements(Connection $connection, Sc
         if ($platform === null) { // @phpstan-ignore-line This is not possible according to doc types, but there is no corresponding type hint in DBAL 2.x
             throw new \RuntimeException('Failed to retrieve Database platform', 1705679147);
         }
+
+        // PostgreSQL maps 'bytea' to 'blob' by default, but our schema uses Types::BINARY
+        // (which also generates BYTEA). Override the mapping so introspected columns match
+        // the desired schema and don't produce persistent diffs.
+        if ($platform instanceof PostgreSQLPlatform) {
+            $platform->registerDoctrineTypeMapping('bytea', 'binary');
+        }
```
Comment (mhsdesign) [inline Neos.ContentRepository.Dbal/Classes/DbalSchemaDiff.php]: This does not seem to have the desired effect already.

```
Subscriptions:
  Neos.Neos:DocumentUriPathProjection:
    Setup: SETUP REQUIRED
      The following SQL statements are required: ALTER TABLE cr_default_p_neos_documenturipath_uri ALTER dimensionspacepointhash SET DEFAULT ''
      ALTER TABLE cr_default_p_neos_documenturipath_uri ALTER dimensionspacepointhash TYPE BYTEA
      ALTER TABLE cr_default_p_neos_documenturipath_uri ALTER origindimensionspacepointhash SET DEFAULT ''
      ALTER TABLE cr_default_p_neos_documenturipath_uri ALTER origindimensionspacepointhash TYPE BYTEA
```

Also it needs to be covered by a test that `cr:status` should have no noise by default;)

File: Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php
Comment (mhsdesign) [inline Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php:1]: granted the change projection does not look that happy anymore with all the `quoteIdentifier` calls but we dont need to worry about the change projection... its only temporary until the content graph figures out what changes are;)

File: Neos.Neos/Tests/Behavior/Features/Bootstrap/RoutingTrait.php [hunk 2]
```diff
@@ -263,6 +263,24 @@ public function tableContainsExactly(TableNode $expectedRows): void
             $this->currentContentRepository->id
         );
         $actualResult = $dbal->fetchAllAssociative('SELECT ' . $columns . ' FROM ' . $tablePrefix . '_uri ORDER BY nodeaggregateidpath, dimensionspacepointhash');
+        $expectedColumnNames = array_keys($expectedRows->getHash()[0]);
+        $actualResult = array_map(static function (array $row) use ($expectedColumnNames) {
```
Comment (mhsdesign) [inline Neos.Neos/Tests/Behavior/Features/Bootstrap/RoutingTrait.php]: whatthathek?:D

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentNodeInfo.php [hunk 3]
```diff
@@ -251,4 +253,19 @@ public function __toString(): string
         return ($this->source['nodeaggregateid'] ?? '<unknown nodeAggregateId>')
             . '@' . ($this->source['dimensionspacepointhash'] ?? '<unkown dimensionSpacePointHash>');
     }
+
+    /**
+     * PostgreSQL returns bytea columns as stream resources, while MariaDB/MySQL returns strings.
+     */
+    private static function binaryToString(mixed $value): string
```
Comment (mhsdesign) [inline Neos.Neos/Classes/FrontendRouting/Projection/DocumentNodeInfo.php]: does this make sense? Regarding database optimisation and performance? Or should Postgres just use string columns?

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -270,16 +312,29 @@ jobs:
...
           # composer test:behavioral
 
           # DEBUG MODE: ALTERNATIVELY, comment in the following lines to dump the DB.
           # do not exit the script if the tests break; as we want to upload the database dumps afterwards.
           set +e
-          composer test:behavioral:stop-on-failure
+
+          # todo introduce dedicated composer scripts to run adapter specific tests
+          # todo like this the retVal does not work correctly and the test pass event if they fail
+          ${{ case(
```
Comment (mhsdesign) [inline .github/workflows/build.yml]: add next-to `composer test:behavioral` a 'composer test:behavioral-postgresql`?

**Review by mhsdesign — COMMENTED** [review]
Some things i like to discuss before merging. Also thanks @anselmpaul for the tremendous work!

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentNodeInfo.php (hunk 3 above)
Comment (anselmpaul) [reply Neos.Neos/Classes/FrontendRouting/Projection/DocumentNodeInfo.php]: changed to string column for postgres in 630c00a

File: Neos.ContentRepository.Dbal/Classes/DbalSchemaDiff.php (hunk 1 above)
Comment (anselmpaul) [reply Neos.ContentRepository.Dbal/Classes/DbalSchemaDiff.php]: removed in a82f1bbc6cb3f20957bf32fbd473369f105ccaf5

Comment (anselmpaul) [reply Neos.ContentRepository.Dbal/Classes/DbalSchemaDiff.php]: and added a test for cr:setup in 1888802dc04f54403b04d2021971a731f9a2a445

**Comment by mhsdesign** [comment]
Hi @anselmpaul ill try to cherry pick your adjustments but feel also free to just create a PR from your fork ontop of this PR and base it onto `bugfix/postgres-compatibility-except-graph`

This way i can simply hit merge;)

EDIT: Rebased now.

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentNodeInfo.php (hunk 3 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/FrontendRouting/Projection/DocumentNodeInfo.php]: I also discussed this with christian and we agree that as long as dbal does not abstract this properly during fetching and we have problems with the schema diff its not worth for the neos projections.

The independent PostgreSQL content graph might still choose to use bytea which might make sense and bring the desired performance as the graph is the biggest projection.

TLTR: For the Neos projections this is currently not relevant, so thanks for reverting this.

File: Neos.Neos/Classes/AssetUsage/Domain/AssetUsageRepository.php [hunk 5]
```diff
@@ -36,23 +34,7 @@ public function __construct(
     public function findUsages(ContentRepositoryId $contentRepositoryId, AssetUsageFilter $filter): AssetUsages
     {
         $queryBuilder = $this->dbal->createQueryBuilder();
-        if ($this->dbal->getDatabasePlatform() instanceof MariaDBPlatform) {
-            $select = '*';
-        } else {
-            $select = 'ANY_VALUE(assetid) as assetid,
```
Comment (mhsdesign) [inline Neos.Neos/Classes/AssetUsage/Domain/AssetUsageRepository.php:42]: note: moved unchanged below to line 79

File: Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php [hunk 6]
```diff
@@ -644,45 +648,58 @@ private function moveNode(
             $removedDelta++;
         }
 
+        $platform = $this->dbal->getDatabasePlatform();
+        $slash = "'" . "/" . "'";
+        // Inline integer offsets directly into SQL to avoid PostgreSQL interpreting
+        // string-typed parameters as regex patterns in SUBSTRING(text, text) overload
```
Comment (mhsdesign) [inline Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php]: this sounds off but somehow the php values indeed become strings in Postgresql 

Using doctrines parameter types does also not change anything

```
[
    'sourceNodeAggregateIdPathOffset' => ParameterType::INTEGER,
    'sourceUriPathOffset' => ParameterType::INTEGER,
]
```

The only way is the cast, but then we can also inline ^^

```
CAST(:sourceNodeAggregateIdPathOffset as integer)
```

> An exception occurred while executing a query: SQLSTATE[25P02]: In failed sql transaction: 7 ERROR:  current transaction is aborted, commands ignored until end of transaction block in

Comment (mhsdesign) [reply Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathProjection.php]: I cleaned up the code and validated that nothing got lost;) Now its pretty _and_ it works;)

File: Neos.Neos/Tests/Behavior/Features/Bootstrap/RoutingTrait.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.Neos/Tests/Behavior/Features/Bootstrap/RoutingTrait.php]: Solved by adjusting the fixtures instead;)

File: Neos.ContentRepository.Dbal/Classes/DbalSchemaFactory.php [hunk 7]
```diff
@@ -101,12 +102,17 @@ public static function columnForDimensionSpacePoint(string $columnName, Abstract
...
     public static function columnForDimensionSpacePointHash(string $columnName, AbstractPlatform $platform): Column
     {
+        if ($platform instanceof PostgreSQLPlatform) {
+            # FIXME PostgreSQL understands BYTEA (a binary stream), but this requires special read handling. Since the hash is
+            # plain ASCII hex, STRING is used on PostgreSQL to keep read/write handling uniform across platforms.
+            return (new Column($columnName, Type::getType(Types::STRING)))
+                ->setLength(32)
+                ->setDefault('');
+        }
         return (new Column($columnName, Type::getType(Types::BINARY)))
```
Comment (kdambekalns) [inline Neos.ContentRepository.Dbal/Classes/DbalSchemaFactory.php:116]: Why is the column `BINARY` in the first place, if this is a plain ASCII string?

**Review by kdambekalns — APPROVED** [review]

**Comment by kdambekalns** [comment]
Leaves only the minimum PostgreSQL version requirement (v16+) to be documented.

File: Neos.ContentRepository.Dbal/Classes/DbalSchemaFactory.php (hunk 7 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Dbal/Classes/DbalSchemaFactory.php:116]: The dimensions space point hash was said to be stored more efficiently in binary. Its a fixed lengh hash `md5(json_encode($identityComponents, JSON_THROW_ON_ERROR))`

That was christians master plan:)
